File: backend/app/routers/ai_reader_queue.py
"""AI reader queue endpoints. Queue disabled (table dropped); extracted-details still works from JSON files."""
from fastapi import APIRouter, HTTPException, Query

from app.config import DEALER_ID, get_ocr_output_dir, get_uploads_dir
from app.services.ocr_service import OcrService

router = APIRouter(prefix="/ai-reader-queue", tags=["ai-reader-queue"])


@router.get("")
def list_ai_reader_queue(limit: int = 200) -> list[dict]:
    """Queue disabled; returns empty list."""
    return []


@router.post("/process-next")
def process_next_ocr() -> None:
    """Queue disabled; returns None."""
    return None


@router.get("/extractions")
def list_extractions(limit: int = 200) -> list[dict]:
    """Queue disabled; returns empty list."""
    return []


@router.get("/extracted-details")
def get_extracted_details(
    subfolder: str,
    dealer_id: int | None = Query(None, description="Dealer ID; uses app default if omitted"),
) -> dict:
    """Return structured vehicle (and customer) details for a subfolder from JSON files.
    May include extraction_error when Aadhar QR failed; vehicle and insurance are still returned."""
    did = dealer_id if dealer_id is not None else DEALER_ID
    service = OcrService(
        uploads_dir=get_uploads_dir(did),
        ocr_output_dir=get_ocr_output_dir(did),
    )
    details = service.get_extracted_details(subfolder)
    if details is None:
        raise HTTPException(status_code=404, detail="No extracted details for this subfolder")
    return details


@router.get("/insurance-extraction")
def get_insurance_extraction(
    subfolder: str,
    dealer_id: int | None = Query(None, description="Dealer ID; uses app default if omitted"),
) -> dict:
    """Debug: return what Textract extracted from Insurance.jpg (raw + parsed) and file status."""
    import re
    from pathlib import Path

    did = dealer_id if dealer_id is not None else DEALER_ID
    ocr_dir = get_ocr_output_dir(did)
    uploads_dir = get_uploads_dir(did)

    def _safe(s: str) -> str:
        return re.sub(r"[^\w\-]", "_", (s or "").strip()) or "default"

    safe = _safe(subfolder)
    subfolder_path = ocr_dir / safe
    upload_path = uploads_dir / subfolder

    result: dict = {
        "subfolder": subfolder,
        "insurance_jpg_exists": (upload_path / "Insurance.jpg").exists(),
        "ocr_files": sorted(p.name for p in subfolder_path.iterdir()) if subfolder_path.exists() else [],
        "insurance_from_details": None,
        "insurance_ocr_json": None,
        "raw_ocr_txt": None,
    }

    service = OcrService(uploads_dir=uploads_dir, ocr_output_dir=ocr_dir)
    details = service.get_extracted_details(subfolder)
    if details and details.get("insurance"):
        result["insurance_from_details"] = details["insurance"]

    insurance_ocr = subfolder_path / "insurance_ocr.json"
    if insurance_ocr.exists():
        try:
            import json
            result["insurance_ocr_json"] = json.loads(insurance_ocr.read_text(encoding="utf-8"))
        except Exception as e:
            result["insurance_ocr_json_error"] = str(e)

    raw_ocr_txt = subfolder_path / "Raw_OCR.txt"
    if raw_ocr_txt.exists():
        result["raw_ocr_txt"] = raw_ocr_txt.read_text(encoding="utf-8")

    insurance_txt = subfolder_path / "Insurance.txt"
    if insurance_txt.exists():
        result["insurance_txt_preview"] = insurance_txt.read_text(encoding="utf-8", errors="replace")[:2000]

    return result


@router.get("/process-status")
def process_status() -> dict:
    """Queue disabled; returns sleeping."""
    return {"status": "sleeping", "processed_count": 0}


@router.post("/empty")
def empty_queue() -> dict:
    """Queue disabled; no-op."""
    return {"ok": True, "deleted": 0}


@router.post("/process-all")
def process_all() -> dict:
    """Queue disabled; no-op."""
    return {"started": False, "message": "AI reader queue is disabled"}


@router.post("/{item_id:int}/reprocess")
def reprocess_item(item_id: int) -> dict:
    """Queue disabled; returns 404."""
    raise HTTPException(status_code=404, detail="AI reader queue is disabled")

File: client/src/api/aiReaderQueue.ts
import { apiFetch, ApiHttpError } from "./client";
import type {
  AiReaderQueueItem,
  ProcessStatusResponse,
  ExtractedDetailsResponse,
} from "../types";

export async function getAiReaderQueue(
  limit = 200
): Promise<AiReaderQueueItem[]> {
  return apiFetch<AiReaderQueueItem[]>(`/ai-reader-queue?limit=${limit}`);
}

export async function getProcessStatus(): Promise<ProcessStatusResponse> {
  return apiFetch<ProcessStatusResponse>("/ai-reader-queue/process-status");
}

export async function startProcessAll(): Promise<{
  started: boolean;
  message: string;
}> {
  return apiFetch<{ started: boolean; message: string }>(
    "/ai-reader-queue/process-all",
    { method: "POST" }
  );
}

export async function emptyAiReaderQueue(): Promise<{ ok: boolean; deleted: number }> {
  return apiFetch<{ ok: boolean; deleted: number }>(
    "/ai-reader-queue/empty",
    { method: "POST" }
  );
}

export async function reprocessQueueItem(
  itemId: number
): Promise<{ ok: boolean; id: number; message: string }> {
  return apiFetch<{ ok: boolean; id: number; message: string }>(
    `/ai-reader-queue/${itemId}/reprocess`,
    { method: "POST" }
  );
}

/** Get structured extracted details (vehicle, customer) for a subfolder. Returns null if not found (e.g. not yet processed). */
export async function getExtractedDetails(
  subfolder: string,
  dealerId?: number
): Promise<ExtractedDetailsResponse | null> {
  const params = new URLSearchParams();
  params.set("subfolder", subfolder);
  if (dealerId != null) params.set("dealer_id", String(dealerId));
  try {
    return await apiFetch<ExtractedDetailsResponse>(
      `/ai-reader-queue/extracted-details?${params.toString()}`
    );
  } catch (e) {
    if (e instanceof ApiHttpError && e.status === 404) return null;
    throw e;
  }
}

/** Debug: get what Textract extracted from Insurance.jpg (raw + parsed) and file status. */
export async function getInsuranceExtraction(subfolder: string): Promise<{
  subfolder: string;
  insurance_jpg_exists: boolean;
  ocr_files: string[];
  insurance_from_details: Record<string, string> | null;
  insurance_ocr_json: Record<string, string> | null;
  raw_ocr_txt: string | null;
  insurance_txt_preview?: string;
}> {
  const params = new URLSearchParams();
  params.set("subfolder", subfolder);
  return apiFetch(`/ai-reader-queue/insurance-extraction?${params.toString()}`);
}
